File: services/server/src/safe_socket/safe_socket.py
import socket


def short_read(socket: socket.socket, size):
    data = b""

    while len(data) < size:
        chunk = socket.recv(size - len(data))

        if chunk == b"":
            raise ConnectionError("socket connection closed")

        data += chunk

    return data

def short_write(socket: socket.socket, data):
    total_sent = 0

    while total_sent < len(data):
        sent = socket.send(data[total_sent:])
        total_sent += sent

    return total_sent

def recv_all(socket: socket.socket, size):
    return short_read(socket, size)

def send_all(socket: socket.socket, bytes):
    return short_write(socket, bytes)

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func ShortWrite(socket io.Writer, bytes []byte) error {
	sent := 0

	for sent < len(bytes) {
		n, err := socket.Write(bytes[sent:])

		if err != nil {
			return err
		}

		sent += n
	}

	return nil
}

func ShortRead(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	received := 0
	for received < size {
		n, err := socket.Read(buff[received:size])
		if err != nil {
			return nil, err
		}
		if n == 0 {
			return nil, io.ErrUnexpectedEOF
		}
		received += n
	}
	return buff, nil
}

func SendAll(socket io.Writer, bytes []byte) error {
	return ShortWrite(socket, bytes)
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	return ShortRead(socket, size)
}
